electrumsv/network.py

```diff
@@ -1027,305 +297,157 @@ async def _main_task(self) -> None:
-
-    async def _maybe_switch_main_server(self, reason: SwitchReason) -> None:
-        now = time.time()
-        max_height = max((session.tip.height for session in self.sessions
-            if session.tip is not None), default=0)
-        for session in self.sessions:
-            if session.tip is not None and session.tip.height > max_height - 2:
-                session.server.state.last_good = now
-        # Give a 60-second breather for a lagging server to catch up
-        good_servers = [session.server for session in self.sessions
-                        if session.server.state.last_good > now - 60]
-        if not good_servers:
-            logger.warning('no good servers available')
-        elif self.main_server not in good_servers:
-            if self.auto_connect():
-                await self._set_main_server(random.choice(good_servers), reason)
-            else:
-                logger.warning("main server %s is not good, but retaining it because "
-                    "auto-connect is off", self.main_server)
-
-    async def _monitor_lagging_sessions(self) -> None:
-        '''Monitor which sessions are lagging.
-
-        If the main server is lagging switch the main server if auto_connect.
-        '''
-        while True:
-            async with ignore_after(20):
-                await self.sessions_changed_event.wait()
-            await self._maybe_switch_main_server(SwitchReason.lagging)
+    def get_local_height(self) -> int:
+        chain = self.chain()
+        # This can be called from network_dialog.py when there is no chain
+        return cast(int, chain.height) if chain else 0
 
-    async def _monitor_wallets(self, group: TaskGroup) -> None:
+    def get_local_tip_hash(self) -> bytes:
+        chain = self.chain()
+        assert chain is not None
+        return cast(bytes, chain.tip.hash)
+
+    async def get_server_tip_async(self) -> Header:
+        """raises `ServiceUnavailableError`"""
+        assert self.headers_client is not None
+        raw_tip = await self.headers_client.get_chain_tips(longest_chain_only=True)
+        raw_header = raw_tip[0:80]
+        height = bitcoinx.le_bytes_to_int(raw_tip[80:84])
+        return Net._net.COIN.deserialized_header(raw_header, height)
+
+    # TODO(post-1.4.0) Need a sound system for banning and selecting a better provider
+    async def _synchronize_initial_headers_async(self) -> None:
+        RETRY_TIMEOUT = 10  # seconds
         while True:
-            job, wallet = await self._wallet_jobs.get()
-            logger.error('unknown wallet job %s', job)
+            tip_obj: Header = await self.get_server_tip_async()
+            self.server_chain_tip = tip_obj
+            while self.get_local_height() < tip_obj.height:
+                heights = [height for height in range(self.get_local_height(),
+                    tip_obj.height + 1)]
+                try:
+                    await self._request_and_connect_headers_at_heights(heights)
```

Comment:
Heights are not meaningful. What is relevant is whether the server tip is known to us. If it is not, then we find out the fork height from a known header in the local store and sync up from there, which would be batch from a base hash. We should do this for each different chain (and it should support multiple servers with the same chain tips and only fetching from one of them rather than all of them). This is what the old code did.

Reply:
The request by height is used for where we locate the fork point.

Reply:
This code already does exactly that AFAICT. `self.get_local_height()` gets the local header height and fetches from that as a base up to the tip of the server. 

If there's a reorg then there will inevitably be a MissingHeader exception raised so it backfills.

The only thing missing is the new request to add in support for multiple servers (which I was going to do in a separate pull request).

I'm not seeing the problem here... sounds like you are just describing what the code does back to me.

Reply:
It is unclear to me if this is the right approach without the multiple server handling being factored in.